pkg/container-collection/options.go

```diff
@@ -252,28 +255,175 @@ func WithPubSub(funcs ...pubsub.FuncNotify) ContainerCollectionOption {
 	}
 }
 
+// getExpectedOwnerReference returns a resource only if it has an expected kind.
+// In the case of multiple references, it first tries to find the controller
+// reference. If there does not exist or it does not have an expected kind, the
+// function will try to find the first resource with one of the expected
+// resource kinds. Otherwise, it returns nil.
+func getExpectedOwnerReference(ownerReferences []metav1.OwnerReference) *metav1.OwnerReference {
+	// From: https://kubernetes.io/docs/concepts/workloads/controllers/
+	expectedResKinds := map[string]struct{}{
+		"Deployment":            {},
+		"ReplicaSet":            {},
+		"StatefulSet":           {},
+		"DaemonSet":             {},
+		"Job":                   {},
+		"CronJob":               {},
+		"ReplicationController": {},
+	}
+
+	var ownerRef *metav1.OwnerReference
+
+	for i, or := range ownerReferences {
+		if _, ok := expectedResKinds[or.Kind]; !ok {
+			continue
+		}
+
+		if or.Controller != nil && *or.Controller {
+			// There is at most one controller reference per resource
+			return &or
+		}
+
+		// Keep track of the first expected reference in case it will be needed
+		if ownerRef == nil {
+			ownerRef = &ownerReferences[i]
+		}
+	}
+
+	return ownerRef
+}
+
+func getKubeClientDynamic() (dynamic.Interface, error) {
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, fmt.Errorf("couldn't get Kubernetes config: %w", err)
+	}
+	dynamicClient, err := dynamic.NewForConfig(config)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't create dynamic Kubernetes client: %w", err)
+	}
+	return dynamicClient, nil
+}
+
+func getOwnerReferences(dynamicClient dynamic.Interface,
+	resNamespace, resKind, resGroupVersion, resName string,
+) ([]metav1.OwnerReference, error) {
+	gv, err := schema.ParseGroupVersion(resGroupVersion)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse %s/%s groupVersion %s: %w",
+			resNamespace, resName, resGroupVersion, err)
+	}
+
+	params := schema.GroupVersionResource{
+		Group:    gv.Group,
+		Version:  gv.Version,
+		Resource: resKind,
+	}
+
+	res, err := dynamicClient.
+		Resource(params).
+		Namespace(resNamespace).
+		Get(context.TODO(), resName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("cannot fetch %s/%s %s/%s: %w",
+			resKind, resGroupVersion, resNamespace, resName, err)
+	}
+
+	return res.GetOwnerReferences(), nil
+}
+
+func ownerReferenceEnrichment(
+	containerDefinition *pb.ContainerDefinition,
+	ownerReferences []metav1.OwnerReference,
+) error {
+	if containerDefinition.OwnerReference != nil {
+		return fmt.Errorf("container %s/%s/%s already has owner reference: %v",
+			containerDefinition.Namespace, containerDefinition.Podname, containerDefinition.Name,
+			containerDefinition.OwnerReference)
```

Comment:
Why to return an error? The idea is to do nothing if the owner reference is already set.